### PR TITLE
Resolve #11825 - no automatic quantity conversion: App::anyToQuantity

### LINKED ISSUE #11825: [Problem] no automatic quantity conversion: `App::anyToQuantity` does not handle `string` case
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

PROBLEM:

currently it is impossible to handle automatic quantity conversion in property expressions
due to:  `App::anyToQuantity` does not handle `string` case
https://github.com/FreeCAD/FreeCAD/blob/8d51d78ce7e44526be6a84daa6dfa2a62c7a1896/src/App/Expression.cpp#L520

VERIFICATION:

* add new document
* add new object (say spreadsheet)
* add new text property: `App::PropertyString` called  `size_text` = `10 mm`
* add new data property: `App::PropertyLength` called  `size_quantity`
* set property expression for the `size_quatity` = `size_text`
* click "recompute"
* observe error message: `<Exception> Failed to convert to Quantity`

<details>
<summary><b>Sample stack trace</b></summary>

```
libFreeCADApp.so!App::anyToQuantity(const App::any & value, const char * msg) (/home/work/source/git/FreeCAD_VS/src/App/Expression.cpp:535)
libFreeCADApp.so!App::isAnyEqual(const App::any & v1, const App::any & v2) (/home/work/source/git/FreeCAD_VS/src/App/Expression.cpp:577)
libFreeCADApp.so!App::PropertyExpressionEngine::execute(App::PropertyExpressionEngine * const this, App::PropertyExpressionEngine::ExecuteOption option, bool * touched) (/home/work/source/git/FreeCAD_VS/src/App/PropertyExpressionEngine.cpp:694)
libFreeCADApp.so!App::Document::_recomputeFeature(App::Document * const this, App::DocumentObject * Feat) (/home/work/source/git/FreeCAD_VS/src/App/Document.cpp:3127)
libFreeCADApp.so!App::Document::recompute(App::Document * const this, const std::vector<App::DocumentObject*, std::allocator<App::DocumentObject*> > & objs, bool force, bool * hasError, int options) (/home/work/source/git/FreeCAD_VS/src/App/Document.cpp:2856)
libFreeCADApp.so!App::DocumentPy::recompute(App::DocumentPy * const this, PyObject * args) (/home/work/source/git/FreeCAD_VS/src/App/DocumentPyImp.cpp:584)
libFreeCADApp.so!App::DocumentPy::staticCallback_recompute(PyObject * self, PyObject * args) (/home/work/source/git/FreeCAD_VS/build/debug/src/App/DocumentPy.cpp:1949)
libpython3.11.so.1.0!cfunction_call(PyObject * func, PyObject * args, PyObject * kwargs) (/home/user0/.cache/debuginfod_client/fa6a19d79d578b6fa7c904274aae9deec2331d44/source##usr##src##debug##python##Python-3.11.6##Objects##methodobject.c:553)
libpython3.11.so.1.0!_PyObject_MakeTpCall(PyThreadState * tstate, PyObject * callable, PyObject * const * args, Py_ssize_t nargs, PyObject * keywords) (/home/user0/.cache/debuginfod_client/fa6a19d79d578b6fa7c904274aae9deec2331d44/source##usr##src##debug##python##Python-3.11.6##Objects##call.c:214)
libpython3.11.so.1.0!_PyEval_EvalFrameDefault(PyThreadState * tstate, _PyInterpreterFrame * frame, int throwflag) (/home/user0/.cache/debuginfod_client/fa6a19d79d578b6fa7c904274aae9deec2331d44/source##usr##src##debug##python##Python-3.11.6##Python##ceval.c:4760)
libpython3.11.so.1.0!_PyEval_EvalFrame(int throwflag, struct _PyInterpreterFrame * frame, PyThreadState * tstate) (/home/user0/.cache/debuginfod_client/fa6a19d79d578b6fa7c904274aae9deec2331d44/source##usr##src##debug##python##Python-3.11.6##.##Include##internal##pycore_ceval.h:73)
libpython3.11.so.1.0!_PyEval_Vector(PyThreadState * tstate, PyFunctionObject * func, PyObject * locals, PyObject * const * args, size_t argcount, PyObject * kwnames) (/home/user0/.cache/debuginfod_client/fa6a19d79d578b6fa7c904274aae9deec2331d44/source##usr##src##debug##python##Python-3.11.6##Python##ceval.c:6425)
libpython3.11.so.1.0!PyEval_EvalCode(PyObject * co, PyObject * globals, PyObject * locals) (/home/user0/.cache/debuginfod_client/fa6a19d79d578b6fa7c904274aae9deec2331d44/source##usr##src##debug##python##Python-3.11.6##Python##ceval.c:1140)
libpython3.11.so.1.0!run_eval_code_obj(PyThreadState * tstate, PyCodeObject * co, PyObject * globals, PyObject * locals) (/home/user0/.cache/debuginfod_client/fa6a19d79d578b6fa7c904274aae9deec2331d44/source##usr##src##debug##python##Python-3.11.6##Python##pythonrun.c:1710)
libpython3.11.so.1.0!run_mod(mod_ty mod, PyObject * filename, PyObject * globals, PyObject * locals, PyCompilerFlags * flags, PyArena * arena) (/home/user0/.cache/debuginfod_client/fa6a19d79d578b6fa7c904274aae9deec2331d44/source##usr##src##debug##python##Python-3.11.6##Python##pythonrun.c:1731)
libpython3.11.so.1.0!PyRun_StringFlags(const char * str, int start, PyObject * globals, PyObject * locals, PyCompilerFlags * flags) (/home/user0/.cache/debuginfod_client/fa6a19d79d578b6fa7c904274aae9deec2331d44/source##usr##src##debug##python##Python-3.11.6##Python##pythonrun.c:1601)
libFreeCADBase.so!Base::InterpreterSingleton::runString[abi:cxx11](char const*)(Base::InterpreterSingleton * const this, const char * sCmd) (/home/work/source/git/FreeCAD_VS/src/Base/Interpreter.cpp:249)
libFreeCADGui.so!Gui::Command::_runCommand(const char * file, int line, Gui::Command::DoCmd_Type eType, const char * sCmd) (/home/work/source/git/FreeCAD_VS/src/Gui/Command.cpp:725)
libFreeCADGui.so!Gui::Command::_doCommand(const char * file, int line, Gui::Command::DoCmd_Type eType, const char * sCmd) (/home/work/source/git/FreeCAD_VS/src/Gui/Command.cpp:677)
libFreeCADGui.so!StdCmdRefresh::activated(StdCmdRefresh * const this, int iMsg) (/home/work/source/git/FreeCAD_VS/src/Gui/CommandDoc.cpp:1483)
```

</details>

Note: this is a contrived example, actual use case is quantity extraction from complicated `App::PropertyPythonObject`

### Full version info

```shell
OS: Arch Linux (KDE/plasma)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.35451 (Git)
Build type: None
Branch: main
Hash: 8d51d78ce7e44526be6a84daa6dfa2a62c7a1896
Python 3.11.6, Qt 5.15.11, Coin 4.0.1, Vtk 9.3.0, OCC 7.7.2
Locale: English/United States (en_US)
```

### Subproject(s) affected?

None

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by Andrei-Pozolotin**
Resolve #11825 - no automatic quantity conversion:` App::anyToQuantity`

### CODE REVIEW

File: src/App/Expression.cpp [hunk 1]
```diff
@@ -517,6 +518,29 @@ Py::Object pyFromQuantity(const Quantity &quantity) {
...
+static const std::regex REGEX_QUANTITY(
+    R"(\s*)"
+    R"(([-+]?[0-9]*[.]?[0-9]+([eE][-+]?[0-9]+)?))" // value
+    R"(\s*)"
+    R"(([^\s]+)?)" // unit
+    R"(\s*)"
+);
+
+// https://github.com/FreeCAD/FreeCAD/issues/11825
+static Quantity parseQuantityFromText(std::string text) {
```
Comment (chennes) [inline src/App/Expression.cpp:530]: Are you familiar with the Google Test framework? I'd love to see a couple of unit tests targeting this method.

**Comment by Andrei-Pozolotin** [comment]
@chennes  : added a couple of unit tests, please review

**Comment by chennes** [comment]
@Andrei-Pozolotin very nice. Can you also add two tests for the two different conditions you are adding to `anyToQuantity` to ensure they are getting called as expected (and to make sure any future changes don't break the new functionality)?

**Comment by Andrei-Pozolotin** [comment]
@chennes  : added a couple of dependency tests, please review

**Comment by chennes** [comment]
Lovely, thank you. I'll wait for the CI to run and then merge it.

**Comment by Andrei-Pozolotin** [comment]
@chennes : I squashed commits, please restart build

**Comment by Andrei-Pozolotin** [comment]
@chennes  : Chris - thank you for taking care of this.
I am testing it now in the context of target plugin to see if anything else comes up.

**Pull request merged**